src/lspspec.sml

```diff
@@ -0,0 +1,452 @@
+structure LspSpec = struct 
+
+  datatype lspError = InternalError of string
+                    | ServerNotInitialized
+  exception LspError of lspError
+
+  fun debug (str: string): unit =
+      (TextIO.output (TextIO.stdErr, str ^ "\n\n"); TextIO.flushOut TextIO.stdErr)
+
+  fun trim (s: substring): substring =
+      Substring.dropr
+        (fn c => c = #" " orelse c = #"\n" orelse c = #"\r")
```

Comment:
Why not replace this hardcoded space-characters check with `Char.isSpace`?   Are you intentionally trying to avoid matching tabs and/or locale-specific space characters?

Reply:
Didn't know about Char.isSpace... changed.